`Development`: Add setup documentation for local continuous integration with Podman (#7756)
ls1intum/Artemis: docs/dev/setup/localci-localvc.rst

.. HINT::
    Unix systems: When running the Artemis container on a Unix system, you will have to give the user running the container permission to access the Docker socket by adding them to the ``docker`` group. You can do this by changing the value of ``services.artemis-app.group_add`` in the ``docker/artemis-dev-local-vc-local-ci-mysql.yml`` file to the group ID of the ``docker`` group on your system. You can find the group ID by running ``getent group docker | cut -d: -f3``. The default value is ``999``.

    Windows: If you want to run the Docker containers locally on Windows, you will have to change the value for the Docker connection URI. You can add ``ARTEMIS_CONTINUOUSINTEGRATION_DOCKERCONNECTIONURI="tcp://host.docker.internal:2375"`` to the environment file, found in ``docker/artemis/config/dev-local-vc-local-ci.env``. This overwrites the default value ``unix:///var/run/docker.sock`` for this property defined in ``src/main/resources/config/application-docker.yml``.

Podman as Docker alternative
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`Podman <https://podman.io/>`_ offers a container runtime that is API-compatible with Docker.

> Rather than having a system-wide socket that runs with administrative permissions, though, Podman allows to create containers with only user permissions.

Reviewer: ```suggestion
Rather than having a system-wide socket that runs with administrative permissions, Podman allows to create containers with only user permissions.
```

In single-user setups this might not be as relevant, but offers additional security in a production environment where the Artemis CI has to execute untrusted student code.

> .. admonition:: Podman is supported on a best effort basis.

Reviewer: Would change this callout to a warning, since it is important to be aware of.

Author: I prefer the current style since it allows for a custom title of the box. Since it is pretty much right at the start of the section before any actual setup instructions, I think the box is visible enough.
A warning in my opinion should be reserved to inform the user that the actions are potentially harmful/dangerous (e.g., potential data loss in a database).

>     We are relying on the API compatibility to provide support but are not actively testing against Podman on a test system or in the continuous integration.

Reviewer 2: It's correct grammar to leave out `the` here, as it does not refer to exactly one specific CI. (Grammarly detected this, too)

    If you notice any issues, feel free to open an issue or pull request so that we can try to fix them.

.. note::

    These setup steps are mostly focused on Linux systems.

>     On mac and Windows, both Docker and Podman run the containers in a small virtual machine anyway.

Reviewer: capitalize Mac in order to be consistent with other parts of the documentation.

    Therefore, there is little technical benefit relevant to Artemis for choosing one over the other in local development setups.

>     If in doubt, we recommend to use Docker, since that solution is most likely to be tested by other Artemis developers.

Reviewer: ```suggestion
    If in doubt, we recommend using Docker, since that solution is most likely to be maintained by Artemis developers.
```

Linux setup
"""""""""""

> Podman itself should be available via your regular package manager.

Reviewer: ```suggestion
Podman itself should be available via your regular package managers.
```

Author: I think singular should be right here: ‘your regular package manager’ is one installed on a specific system, that usually only has one.

> After installation, you have to ensure that your user is allowed to create containers.

Reviewer: ```suggestion
After the installation, you have to ensure that your user is allowed to create containers.
```

This is managed by the files ``/etc/subuid`` and ``/etc/subgid``.
Ensure both files contain a line starting with your username.

> If not, you can generate the relevant lines using

Reviewer: ```suggestion
If not, you can generate the relevant lines by executing the following command:
```

.. code-block:: bash

    #! /usr/bin/env sh

    printf "%s:%d:65536\n" "$USER" "$(( $(id -u) * 65536 ))" | tee -a /etc/subuid /etc/subgid

After that, enable the Podman user socket that provides the API for the container management:

.. code-block:: bash

    systemctl --user enable --now podman.socket

Configure the connection to this socket in Artemis by replacing ``${UID}`` with your actual user id (``id -u``):

.. code-block:: yaml

    artemis:
        continuous-integration:
            docker-connection-uri: "unix:///run/user/${UID}/podman/podman.sock"
            # alternatively, if you use the `DOCKER_HOST` environment variable already
            # to tell other tools to use the Podman socket instead of the Docker one:
            # docker-connection-uri: "${DOCKER_HOST}"

Windows or Mac setup
""""""""""""""""""""

Podman offers a `desktop application <https://podman-desktop.io/>`_ application similar to Docker desktop and `CLI tools <https://podman.io>`_ for Windows, macOS, and Linux.

> As with Docker, to run containers on Windows or macOS the runtime has to start a small virtual Linux machine that then actually runs the containers.

Reviewer 2: Please add a comma:
`or macOS`, `the runtime has to`

You can probably connect to this VM similarly as described in the regular setup steps above
(`additional Podman documentation <https://podman-desktop.io/docs/migrating-from-docker/using-the-docker_host-environment-variable>`_).

> .. note::

Reviewer: Not sure if this fits here, but it feels a bit out of place ot me. Maybe move to a preface at the start of the whole section? Maybe others have better ideas!
